compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEUnrollAndDistributeWorkgroup.cpp

```diff
@@ -0,0 +1,503 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "aie/Dialect/AIE/IR/AIEDialect.h"
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "iree-amd-aie/Transforms/Transforms.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-amdaie-unroll-and-distribute-workgroup"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+/// Convert scf.forall ops within a workgroup to scf.for ops
+/// TODO(jornt): use upstream `forallToFor` function once merged.
+LogicalResult workgroupForallToFor(mlir::ModuleOp moduleOp) {
+  IRRewriter rewriter(moduleOp.getContext());
+  WalkResult res = moduleOp->walk([&](AMDAIE::WorkgroupOp workgroupOp) {
+    WalkResult workgroupRes = workgroupOp->walk([&](scf::ForallOp forallOp) {
+      if (failed(scf::forallToForLoop(rewriter, forallOp))) {
+        workgroupOp.emitOpError()
+            << "failed to transform scf.forall to scf.for";
+        return WalkResult::interrupt();
+      }
+      return WalkResult::advance();
+    });
+    if (workgroupRes.wasInterrupted()) return WalkResult::interrupt();
+    return WalkResult::advance();
+  });
+  if (res.wasInterrupted()) return failure();
+  return success();
+}
+
+/// Hoist an affine apply op on a scf.for op's induction variable into that
+/// scf.for block.
+LogicalResult hoistAffineApplyDependingOnFor(mlir::ModuleOp moduleOp) {
+  IRRewriter rewriter(moduleOp.getContext());
+  moduleOp->walk([&](affine::AffineApplyOp applyOp) {
+    (void)hoistForAffineApplyOp(rewriter, applyOp);
+  });
+  return success();
+}
+
+/// Unroll the scf.forall loops selected for parallel execution on the AIE
+/// array. Try to hoist dma ops that don't depend on the loops\ induction
```

Comment:
I might be wrong but did you mean `loops'` instead of `loops\`?

Reply:
Yes, I did. Thanks for the catch.